Add a faster code path for PETSc vectors.

drwells opened the pull request:
This patch lowers the total number of calls to `VecSetValues` in `ConstraintMatrix::distribute_local_to_global` by writing new vector values to a temporary array before sending them off to PETSc.

I used step-40 to check that the performance improved. Here is the (filtered) output with this patch:
```
[drwells@archway step-40]$ make run | grep assembly
| assembly                        |         1 |   0.00966s |       8.7% |
| assembly                        |         1 |    0.0122s |       6.4% |
| assembly                        |         1 |    0.0217s |       5.9% |
| assembly                        |         1 |    0.0431s |       5.6% |
| assembly                        |         1 |    0.0848s |       5.6% |
| assembly                        |         1 |     0.173s |       5.6% |
| assembly                        |         1 |     0.351s |       5.5% |
| assembly                        |         1 |     0.729s |       5.5% |
```

and on master:
```
[drwells@archway step-40]$ make run | grep assembly
| assembly                        |         1 |   0.00672s |       8.9% |
| assembly                        |         1 |    0.0133s |       7.2% |
| assembly                        |         1 |    0.0258s |       6.9% |
| assembly                        |         1 |    0.0513s |       6.8% |
| assembly                        |         1 |     0.101s |       6.6% |
| assembly                        |         1 |     0.203s |       6.5% |
| assembly                        |         1 |     0.414s |       6.4% |
| assembly                        |         1 |      0.93s |       6.9% |
```

so it appears that the whole assembly process takes about 10% less time: a nice little improvement.

That being said, this patch is pretty ugly and I would be happy to hear if someone has an idea for how to make it look less like a hack.

bangerth commented on include/deal.II/lac/constraint_matrix.templates.h:
```diff
@@ -2269,8 +2274,13 @@ ConstraintMatrix::distribute_local_to_global (
+  std::vector<size_type> &cols          = scratch_data->columns;
+  std::vector<number>    &vals          = scratch_data->values;
+  // create arrays for writing into the vector as well
+  std::vector<size_type> &rows          = scratch_data->vector_indices;
```
`rows` suggests that these are matrix rows (because we already have `cols`). Maybe rename `vector_elements` or similar?

bangerth replied:
Or `vector_indices` :-)

bangerth replied:
And on a completely different tangent, I don't find the use of aliases useful because it makes it so much more difficult to figure out who owns the object, what locking semantics are, etc. But that is for a separate patch.

bangerth commented on include/deal.II/lac/constraint_matrix.templates.h:
```diff
@@ -2319,10 +2330,42 @@ ConstraintMatrix::distribute_local_to_global (
+  if (petsc_global_vector != 0 && types_are_equal<PetscScalar, number>::value)
+    {
+      rows.resize(n_local_rows);
+      vector_values.resize(n_local_rows);
```
Can you add a comment before these lines of the kind "Drop elements of the 'rows' and 'vector_values' objects that were not used because we don't write zero elements" to make it clear that you're resizing to a smaller size, dropping unused elements.

bangerth commented on include/deal.II/lac/constraint_matrix.templates.h:
```diff
@@ -2319,10 +2330,42 @@ ConstraintMatrix::distribute_local_to_global (
+#ifdef DEAL_II_WITH_PETSC
+  PETScWrappers::VectorBase *petsc_global_vector
+    = dynamic_cast<PETScWrappers::VectorBase *>(&global_vector);
+  if (petsc_global_vector != 0 && types_are_equal<PetscScalar, number>::value)
```
You won't need the `petsc_global_vector` pointer outside the `if` statement. Maybe write this as
```
  if (PETScWrappers::VectorBase *petsc_global_vector
￼     = (types_are_equal<PetscScalar, number>::value ?
         dynamic_cast<PETScWrappers::VectorBase *>(&global_vector) :
         NULL))
```

bangerth commented on include/deal.II/lac/constraint_matrix.templates.h:
```diff
@@ -2319,10 +2330,42 @@ ConstraintMatrix::distribute_local_to_global (
+      for (size_type row_n=0; row_n<n_local_rows; ++row_n)
+        {
+          global_vector(rows[row_n]) +=
+            static_cast<typename VectorType::value_type>(vector_values[row_n]);
```
Out of curiosity: Don't the other vector classes have an `add` function as well?

drwells commented:
It appears that some of the other vector classes have `add` (`Vector`, `TrilinosWrappers::MPI::Vector` both do) but our own `LinearAlgebra::distributed::Vector` does not. I suspect Trilinos vectors would also benefit from the same optimization: I'll try to generalize things.

drwells commented:
I addressed the comments.

It looks like all we need to do to get rid of the PETSc specific bits is implement `add` for the `la_vector` classes, which I did. I marked this as WIP since those methods need tests. Is there a general design document describing what we expect what methods all vectors (PETSc, deal.II, etc) to implement? I don't know much about these classes.

drwells commented:
I spent some time reading through #745 and related issues. Based on future plans to move things to inherit from `ReadWriteVector`, is it acceptable, for the sake of this patch, to implement `LinearAlgebra::Distributed::Vector::add` and `LinearAlgebra::Distributed::BlockVector::add` where I have? I believe that the answer is yes but I want to bring this up for explicit discussion.

kronbichler commented:
> is it acceptable, for the sake of this patch, to implement LinearAlgebra::distributed::Vector::add and LinearAlgebra::distributed::BlockVector::add where I have? I believe that the answer is yes but I want to bring this up for explicit discussion.

I think this looks reasonable and do not see problems. Let's wait if @Rombur has comments, but in general the design of the vector classes was to start with a rather small interface and extend as necessary. This looks like a good approach of doing so. (Even though my codes do not have these kinds of code in the hot path so I cannot comment on possible alternatives - I guess for PETSc this is the way to go.)

bangerth commented:
I didn't expect you to add these functions to the other vector classes, but I appreciate you did it anyway!

I'm in favor of merging as is. But Since @kronbichler wanted to see what @Rombur has to say, let's wait a couple of days to let him chime in.

drwells commented:
I would still like to add a test for `distributed::BlockVector::add` and to rebase away my mistakes before we merge this.

Rombur commented:
I think that this patch is fine but I don't think it should be extended to the new vector interface, i.e., it should not become part of the `VectorSpaceVector` interface. The `add(indices, values)` should be replaced by creating a  `ReadWriteVector rwv(indices)`, then filling the `ReadWriteVector`, and finally `vector.import(rwv, VectorOperation::add)`.

drwells commented:
I cleaned things up and added the missing test.

@Rombur would you prefer that we do this interface change now or later? I am not sure that I understand what you are requesting.

Rombur commented:
@drwells I am fine with the PR but I am against changing the interface in the new vector, for now. My message was more a reply to @kronbichler about extending the new vector interface.

kronbichler commented on tests/mpi/parallel_vector_19.cc:
```diff
@@ -0,0 +1,94 @@
+// ---------------------------------------------------------------------
+
+
+// use same partition as in prarallel_vector_06 to check add() with std::vector.
```
typo: prarallel_vector_06.

kronbichler commented on tests/mpi/parallel_block_vector_04.cc:
```diff
@@ -0,0 +1,199 @@
+// check global reduction operations (norms, operator==, operator!=) with the
+// parallel block vector class. This is the same as parallel_vector_01 except
+// that we use the new 'add' function to set the values in the block vector
+// (parallel_vector_01 uses vector assignment).
```
please be more explicit when describing the way elements are added: I would prefer 'we add several elements coming from an std::vector to the block vector'. In particular, I would avoid 'new' in this context... ;-)

drwells commented:
@kronbichler I rewrote both descriptions: they now both explicitly describe the `add` method they use.
